### DISCUSSION TITLE
Avalonia macOS App with Xamarin.Mac SDK: AppKitThreadAccessException when accessing NSApplication.SharedWindow.MainWindow

### DISCUSSION THREAD

**Discussion by Sathiyaraman-M** [AvaloniaUI/Avalonia#13162] [Q&A]
I am developing a Avalonia macOS application, using `net7.0-macos` TFM. Currently I am trying to enter full-screen mode with custom `NSApplicationPresentationOptions`

This is the snippet I am trying to run:

```
public void DisableExternalKeyShortcuts()
{
        var window = NSApplication.SharedApplication.MainWindow;
        const NSApplicationPresentationOptions options = NSApplicationPresentationOptions.HideDock
                                                                 | NSApplicationPresentationOptions.HideMenuBar 
                                                                 | NSApplicationPresentationOptions.DisableAppleMenu
                                                                 | NSApplicationPresentationOptions.DisableProcessSwitching
                                                                 | NSApplicationPresentationOptions
                                                                     .DisableSessionTermination
                                                                 | NSApplicationPresentationOptions.DisableHideApplication
                                                                 | NSApplicationPresentationOptions.AutoHideToolbar;
        var optionsDictionary = new NSDictionary<NSString, NSObject>(
                    new NSString("NSFullScreenModeApplicationPresentationOptions"), new NSNumber((long)options));
        window.ContentView!.EnterFullscreenModeWithOptions(NSScreen.MainScreen, optionsDictionary);
}
```

But I am getting `AppKitThreadAccessException: AppKit Consistency error: you are calling a method that can only be invoked from the UI thread.`

I tried wrapping the method content inside `Dispatcher.UIThread.InvokeAsync()` and `DispatchQueue.MainQueue.DispatchAsync()`, but no avail. Besides I didn't fully understand the threading concepts here.

**Comment by thevortexcloud** [comment]
I don't know much about programming for Macs, but it almost sounds like you need to be on the AppKit UI thread rather than Avalonia's?

**Reply by Sathiyaraman-M** [reply]
Yeah I do feel the same that I mostly need the AppKit's UI thread but I don't know how exactly to accomplish that. I tried using the `DispatchQueue` so that I could capture the AppKit's UI thread but no avail. Besides I had used a temporary fix for this using `NSApplication.CheckForIllegalCrossThreadCalls = true`, but the application started to have frequent crashes at startup.

**Comment by Jeric-X** [comment]
Did you call `NSApplication.Init()` before operating the NSWindow? I met the same AppKitThreadAccessException until I call that at the first line of the main method.

**Reply by kevinoneill** [reply]
@Jeric-X you're a scholar and a gentleman ... this resolved accessing the MainWindow for me.
